### PR TITLE
mgr: injectargs support for mgr daemon

### PR DESCRIPTION
**Pull request by vumrao**
mgr: injectargs support for mgr daemon

Fixes: http://tracker.ceph.com/issues/39440
Signed-off-by: Vikhyat Umrao <vumrao@redhat.com>

### CODE REVIEW

**Comment by vumrao** [comment]
- Tested with vstart cluster works fine.

~~~
$ bin/ceph -s                                                         
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-04-24 10:13:22.492 7f0a86e84700 -1 WARNING: all dangerous and experimental features are enabled.
2019-04-24 10:13:22.551 7f0a86e84700 -1 WARNING: all dangerous and experimental features are enabled.
  cluster:
    id:     041c60f3-8778-4b14-94fe-9096a294fe41
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum a,b,c (age 2m)
    mgr: x(active, since 2m)
    mds: a:1 {0=a=up:active} 2 up:standby
    osd: 3 osds: 3 up (since 89s), 3 in (since 89s)
    rgw: 1 daemon active (rgw)
 
  data:
    pools:   6 pools, 56 pgs
    objects: 223 objects, 6.0 KiB
    usage:   6.0 GiB used, 27 GiB / 33 GiB avail
    pgs:     56 active+clean

$ bin/ceph daemon mgr.x config show | grep mgr_stats_period           
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
    "mgr_stats_period": "5",

$ bin/ceph tell mgr.x injectargs '--mgr_stats_period 10'       
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-04-24 10:13:56.125 7f2ee704b700 -1 WARNING: all dangerous and experimental features are enabled.
2019-04-24 10:13:56.218 7f2ee704b700 -1 WARNING: all dangerous and experimental features are enabled.
mgr_stats_period = '10' 

$ bin/ceph daemon mgr.x config show | grep mgr_stats_period           
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
    "mgr_stats_period": "10",


$ bin/ceph daemon mgr.x config show | grep mon_pg_warn_max_object_skew
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
    "mon_pg_warn_max_object_skew": "10.000000",

$ bin/ceph tell mgr.x injectargs '--mon_pg_warn_max_object_skew 20'   
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-04-24 10:14:48.473 7f4f65d38700 -1 WARNING: all dangerous and experimental features are enabled.
2019-04-24 10:14:48.551 7f4f65d38700 -1 WARNING: all dangerous and experimental features are enabled.
mon_pg_warn_max_object_skew = '20.000000' (not observed, change may require restart) 

$ bin/ceph daemon mgr.x config show | grep mon_pg_warn_max_object_skew
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
    "mon_pg_warn_max_object_skew": "20.000000",
~~~

**Comment by vumrao** [comment]
cc @neha-ojha

File: src/mgr/DaemonServer.cc [hunk 1]
```diff
@@ -927,6 +927,22 @@ bool DaemonServer::_handle_command(
     return true;
   }
 
+  if (prefix == "injectargs") {
+    vector<string> argsvec;
+    cmd_getval(cct, cmdctx->cmdmap, "injected_args", argsvec);
+
+    if (argsvec.empty()) {
```
Comment (neha-ojha) [inline src/mgr/DaemonServer.cc]: Looks like we are returning `r = -EINVAL` when empty elsewhere - should keep it the same here

File: src/mgr/MgrCommands.h [hunk 2]
```diff
@@ -208,3 +208,7 @@ COMMAND("device set-life-expectancy name=devid,type=CephString "\
 COMMAND("device rm-life-expectancy name=devid,type=CephString",
 	"Clear predicted device life expectancy",
 	"mgr", "rw")
+COMMAND("injectargs " \
+        "name=injected_args,type=CephString,n=N",
+        "inject configuration arguments into running MGR",
+        "mgr", "rw")
```
Comment (liewegas) [inline src/mgr/MgrCommands.h]: we don't want to list this here, since this isn't a top level CLI command ('ceph injectargs ...').

**Comment by liewegas** [comment]
TBH I don't think we have a nice way to add 'tell' commands to the mgr right now.  But I don't think we really need to. If you want to set arguments like this, you can do ``ceph config set mgr foo bar``.

**Comment by liewegas** [comment]
We are overdue for a general cleanup of the 'ceph tell' and 'ceph daemon ...' interfaces (ideally to unify them) but there are a bunch of annoying fixes/cleanups to make for that to work (mainly separating things out into different message types to distinguish them from CLI commands).

File: src/mgr/DaemonServer.cc (hunk 1 above)
Comment (vumrao) [reply src/mgr/DaemonServer.cc]: Thanks Neha. Sending updated one. Will align how it is handled in mgr.

**Comment by vumrao** [comment]
> We are overdue for a general cleanup of the 'ceph tell' and 'ceph daemon ...' interfaces (ideally to unify them) but there are a bunch of annoying fixes/cleanups to make for that to work (mainly separating things out into different message types to distinguish them from CLI commands).

@liewegas thank you, Sage. I tested the config set and it works perfectly fine for mgr daemon looks like it was implemented for mgr daemon as it is not taking daemon as arg for ceph.conf options. 

~~~
# ceph config set mgr mon_pg_warn_max_object_skew 10
Invalid command: unused arguments: [u'10']
config set <key> <value> :  Set a configuration option at runtime (not persistent)
Error EINVAL: invalid command
~~~

This works:

~~~
# ceph config set mon_pg_warn_max_object_skew 20
Set mon_pg_warn_max_object_skew to 20

# ceph daemon mgr.`hostname -s` config show | grep mon_pg_warn_max_object_skew
    "mon_pg_warn_max_object_skew": "20.000000",
~~~

I think we can close this one as we have planned cleanup for ceph tell and ceph daemon and we have already two ways 2 set one config set and one from daemon config set in the meantime.

**Pull request closed without merging**